### ISSUE TITLE
Revoking subscription authorization on long-lived connections

### ISSUE THREAD

**Issue by abastardi** [slact/nchan#298]
After a given request is authorized and Nchan establishes the subscription, suppose the user's authorization for the subscription needs to be revoked (e.g., the user's role changes to one that should not have access to the subscription). If using a long-lived connection, such as a websocket or EventSource subscriber, the user would continue to receive publications for as long as the connection remains alive. Is there any mechanism to handle this situation? For example, the ability to associate a given connection or subscription with a given user and issue a command to unsubscribe or kill the connection? Or at least some way to force each connection to close and reconnect periodically (and therefore re-authenticate), which would at least limit the time period during which a user could receive publications after losing authorization?

**Comment by slact** [comment]
There is no explicit mechanism to do this, but it's easily accomplished with channel multiplexing. Consider this configuration:

```nginx
location ~ /sub/(\w+) {
  nchan_subscriber;
  nchan_channel_id shared_$1 user_$arg_userid;
  nchan_authorize_request /authorize;
}

location /authorize {
   internal;
  // [...]
}

location ~ /pub/(\w+) {
  //publish to non-user-channels channels
  nchan_publisher;
  nchan_channel_id chan_$1;
}

location /pub/user {
  nchan_publisher;
  nchan_channel_id user_$arg_userid;
}
```

With this config, subscribers send a request to `/sub/mychannel?userid=foo`, which subscribes them to two channnels: `chan_mychannel` and `user_foo`.  You can then publissh to `chan_*` by POSTing to `/pub/...` as usual.
When you want to kick out a particular subscriber, send a DELETE to `/pub/user?userid=id`.

Essentially, in this setup every subscriber subscribes to a shard channel and its own user channel. DELETing a user's channel disconnects that user.

**Comment by abastardi** [comment]
Brilliant. This might be worth a brief mention in the docs, maybe in the "Securing Channels" section of the Details page. It might also be worth explicitly noting that deleting a single channel will disconnect even subscribers who are also subscribed to other channels via multiplexing, requiring them to reconnect to the non-deleted channels. Would you be interested in pull requests to https://github.com/slact/nchapp?

Thank you.

**Comment by slact** [comment]
Yes, it might make sense to document this. You're welcome to submit a pull request to nchapp.

**Comment by abastardi** [comment]
The following is a bit unclear:

> DELETE requests to a multiplexed channel broadcast the deletion to each of the channels it multiplexes, deletes all their messages and kicks out all clients subscribed to any of the channel ids.

Let's say subscriber A is subscribed to channels X and Y via multiplexing, and subscriber B is subscribed to channels Y and Z via multiplexing. If channel X is deleted, subscriber A's connection is closed, so subscriber A loses subscriptions to both X and Y, correct? But in this case, nothing happens to the Y channel or to subscriber B's connections or subscriptions, right? The above sentence makes it sound like because X is multiplexed with Y within one subscriber's connection, deleting channel X will also delete channel Y and therefore disconnect subscriber B.

**Comment by slact** [comment]
Ah, right. This is confusing because I've made no distinction here between *publishing* a DELETE to to a multiplexed location, and how publishing a   to a single channel is handled for multiplexed subscribers.  I'll find a way to explain both cases. If you have a suggestion, please feel free to send a PR. The main page is built from the [README](/slact/nchan/blob/master/README.md), which in turn is generated in part from [nchan_commands.rb](/slact/nchan/blob/master/src/nchan_commands.rb).

**Comment by abastardi** [comment]
Happy to take a shot at it. Let me make sure I understand:

If you have:

```
location ~ /multipub/(\w+)/(\w+)$ {
    nchan_publisher;
    nchan_channel_id "$1" "$2" "another_channel";
  }
```
and send `DELETE` to `/multipub/foo/bar`, it will delete channels foo, bar, and another_channel as well as disconnect any subscribers who were subscribed to any of those channels (whether they were subscribed via multiplexing or not).

If you have:

```
location ~ /sub/channel_A {
    nchan_subscriber;
    nchan_channel_id "channel_A";
  }

location ~ /multisub {
    nchan_subscriber;
    nchan_channel_id "channel_A" "channel_B";
  }

location = /pub/channel_A {
    nchan_publisher;
    nchan_channel_id "channel_A";
  }
```
and send `DELETE` to `/pub/channel_B`, it will delete channel_B and disconnect any subscribers, including subscribers who were subscribed to both channel_A and channel_B via `/multisub`, though it will not affect subscribers who were subscribed to channel_A alone via `/sub/channel_A`.

Is all that correct?

As an aside, would you say multiplexing subscriptions is less useful over HTTP/2, given that separate subscription requests would end up sharing the same TCP connection anyway?

**Comment by slact** [comment]
Yep, you got that right. Basically, publishing a DELETE to any channel kick all its subscribers (and deletes stored messages). Some of the subscribers may be in that deleted channel through a multiplexed subscriber location. Those get kicked out too.

But also, you can publish a DELETE to several channels at once using multiplexed publisher locations.

And to your aside, yes -- well, sort of. Keeping the TCP connection count low stops being an issue with http2, but it's still conceptually useful to multiplex connections within a single subsciber request -- as your use case makes clear.
